include/seastar/core/sharded.hh

```diff
@@ -180,14 +180,26 @@ class sharded {
     struct entry {
         shared_ptr<Service> service;
         promise<> freed;
+
+        explicit entry(shared_ptr<Service> s) noexcept : service(std::move(s)) { }
+
+        void service_deleted() noexcept {
+            freed.set_value();
+        }
+
+        void track_deletion(std::false_type) noexcept {
+            // do not wait for instance to be deleted since it is not going to notify us
+            service_deleted();
+        }
+
+        void track_deletion(std::true_type) noexcept {
+            service->_delete_cb = std::bind(std::mem_fn(&entry::service_deleted), this);
+        }
     };
-    std::vector<entry> _instances;
+    std::vector<std::unique_ptr<entry>> _instances;
```

Comment:
It's really minor, but I can't like the double indirection (unique_ptr->shared_ptr).

How about this:
 1. move `freed` into `async_sharded_service` (which is allocated on the shard)
 2. use `constexpr if` or other tricks to avoid referring to `freed` when `Service` isn't an `async_sharded_service`. After all, in those cases deletion is synchronous.

Reply:
I'll submit this patch separately since it's worthwhile regardless of the way we handle https://github.com/scylladb/seastar/issues/1563

Reply:
> I'll submit this patch separately since it's worthwhile regardless of the way we handle https://github.com/scylladb/seastar/issues/1563

Sorry, I've given it another go and sent a new version including the revised version of this patch.
If this version is not agreed upon I'll send the first patch independently.

Reply:
Please send it independently regardless, it's easier to review it in its own context.